### ISSUE TITLE
Non-deterministic build failures installing npm dependencies

### ISSUE THREAD

**Issue by jessestuart** [Shippable/support#1895]
We're experiencing intermittent build failures due to an error occurring while installing npm dependencies. The specific command is `sudo npm install -g karma karma-cli bower` (the project itself is JVM based, not node based, so the recommendation in #235 did not work). Here's an excerpt from the stacktrace:

```
npm ERR! Linux 3.13.0-57-generic
npm ERR! argv "node" "/usr/bin/npm" "install" "-g" "karma" "karma-cli" "bower"
npm ERR! node v0.10.33
npm ERR! npm  v3.3.3

npm ERR! Maximum call stack size exceeded
```

What's odd is that this occurs on branches that have not changed, and for which builds were passing last week.

I tried retriggering builds with the `[reset_minion]` flag, but got the same result.

**Comment by jawaid** [comment]
@jessestuart  We will look into it and get back to you.

**Comment by jessestuart** [comment]
@jawaid Update: I'm not sure what to make of this, but it seems that part of the problem may be that we're trying to always pull the latest version of `npm` before installing dependencies, i.e., by including `sudo npm install -g npm@latest` in our `before_install` block. 

By setting this to a fixed version (e.g., `2.11.0`) we're able to get builds to go through, at least for the moment. However, I'd still like to understand what's going on here to prevent future problems!

**Comment by trriplejay** [comment]
Hi @jessestuart, I was just going to recommend exactly what you did.

I noticed that you're using npm@latest in your before_install section.  It seems that last week, latest npm was changed from 2.14.3 to 3.3.3, and if you compare your last successful build to your first failed build, you'll see that the installed npm version in that step has changed.

We had at least one other customer experience a slightly different bug with the same root cause, so I think the key is to be specific about which npm version you want to use.
